src/Generator.cpp

```diff
@@ -97,29 +99,551 @@ Outputs compute_outputs(const Target &target,
+    indent--;
+    stream << ind() << "};\n";
+    stream << "\n";
+
+    indent--;
+    stream << ind() << "};\n";
+    stream << "\n";
+
+    for (int i = (int)namespaces.size() - 1; i >= 0 ; --i) {
+        stream << ind() << "}  // namespace " << namespaces[i] << "\n";
+    }
+    stream << "\n";
+
+    stream << ind() << "#endif  // " << guard.str() << "\n";
+}
+
+void verify_same_funcs(Func a, Func b) {
+    user_assert(a.function().get_contents().same_as(b.function().get_contents())) 
+        << "Expected Func " << a.name() << " and " << b.name() << " to match.\n";
+}
+
+void verify_same_funcs(const std::vector<Func>& a, const std::vector<Func>& b) {
+    user_assert(a.size() == b.size()) << "Mismatch in Function vector length.\n";
+    for (size_t i = 0; i < a.size(); ++i) {
+        verify_same_funcs(a[i], b[i]);
+    }
+}
+
+const std::map<std::string, Type> &get_halide_type_enum_map() {
+    static const std::map<std::string, Type> halide_type_enum_map{
+        {"bool", Bool()},
+        {"int8", Int(8)},
+        {"int16", Int(16)},
+        {"int32", Int(32)},
+        {"uint8", UInt(8)},
+        {"uint16", UInt(16)},
+        {"uint32", UInt(32)},
+        {"float32", Float(32)},
+        {"float64", Float(64)}
     };
     return halide_type_enum_map;
 }
 
+std::string halide_type_to_c_source(const Type &t) {
+    static const std::map<halide_type_code_t, std::string> m = {
+        { halide_type_int, "Int" },
+        { halide_type_uint, "UInt" },
+        { halide_type_float, "Float" },
+        { halide_type_handle, "Handle" },
+    };
+    std::ostringstream oss;
+    oss << "Halide::" << m.at(t.code()) << "(" << t.bits() << + ")";
+    return oss.str();
+}
+
+std::string halide_type_to_c_type(const Type &t) {
+    auto encode = [](const Type &t) -> int { return t.code() << 16 | t.bits(); };
+    static const std::map<int, std::string> m = {
+        { encode(Int(8)), "int8_t" },
+        { encode(Int(16)), "int16_t" },
+        { encode(Int(32)), "int32_t" },
+        { encode(Int(64)), "int64_t" },
+        { encode(UInt(1)), "bool" },
+        { encode(UInt(8)), "uint8_t" },
+        { encode(UInt(16)), "uint16_t" },
+        { encode(UInt(32)), "uint32_t" },
+        { encode(UInt(64)), "uint64_t" },
+        { encode(Float(32)), "float" },
+        { encode(Float(64)), "double" },
+        { encode(Handle(64)), "void*" }
+    };
+    return m.at(encode(t));
+}
+
+LoopLevel get_halide_undefined_looplevel() {
+    static LoopLevel undefined(Func("__undefined_looplevel_func"), Var("__undefined_looplevel_var"));
+    return undefined;
+}
+
+const std::map<std::string, LoopLevel> &get_halide_looplevel_enum_map() {
+    static const std::map<std::string, LoopLevel> halide_looplevel_enum_map{
+        {"root", LoopLevel::root()},
+        {"undefined", get_halide_undefined_looplevel()},
+        {"inline", LoopLevel()},
```

Comment:
The textual values specified here can only be used in build rules (e.g. Makefiles); since there is no unique textual representation for a Func, there can't be a unique textual representation for an arbitrary LoopLevel. That said: these three possibilties are "special" and arguably worth providing a way to specify at build time.